### PR TITLE
Dhaval/merging all models with restructure model

### PR DESCRIPTION
**Pull request by d-val**
This PR contains:
- Implementation of next word prediction with `gpt2_xl` using test function `test_next_word_gpt2_xl`
- Implementation of BERT model testing using model `bert-base-uncased` within function `test_next_word_mask_bert_base_uncased`
- Moved next `huggingface.py`'s word prediction and neural recordings outside of `digest_text` into it's own function `predict_next_word`.

Removed (at least for this PR):
- ~~Implementation of `fill_mask` to be used for models in the BERT family (which cannot be really used for next word prediction, but are natural candidates for word mask filling)~~ ---> Removed because @mschrimpf advises to focus on next word prediction for now. 
- ~~Implementation of Google's `t5-small` model instead of the large `t5-11b` model because the latter model's file is 43 GB from HuggingFace and requires 90 GB of RAM to run (5-10min to get one round of debugging to be tested). `t5-11b` model  is just a larger version of the same model/API/definitions. Tested using test function `test_fill_mask_t511b_and_representation`.~~ --> Removed as it breaks the model interface to have to feed in, in addition to stimuli text, either `labels` or `decoder_input_ids`. 
- ~~Implementation of `input_target_sequence` task required for the `T5` family of models.~~ --> Removed as it breaks the model interface to have to feed in, in addition to stimuli text, either `labels` or `decoder_input_ids`.  

Note :
1. As discussed, this works as per the screenshot below 
Screen Shot 2022-07-29 at 17 04 26

### CODE REVIEW

File: brainscore_language/models/huggingface.py [hunk 1]
```diff
@@ -34,18 +34,24 @@ def __init__(
...
+        self.task_function_mapping_dict = {
+            ArtificialSubject.Task.next_word: self.predict_next_word,
+        }
+
     def identifier(self):
         return self.model_id
 
     def perform_behavioral_task(self, task: ArtificialSubject.Task):
         self.behavioral_task = task
+        self.run_behavioral_experiment = self.task_function_mapping_dict[task]
```
Comment (mschrimpf) [inline brainscore_language/models/huggingface.py:46]: Since this function doesn't really "run" anything, but rather converts the output to something behaviorally meaningful, I propose renaming

```suggestion
        self.output_to_behavior = self.task_function_mapping_dict[task]
```

File: brainscore_language/models/huggingface.py [hunk 2]
```diff
@@ -77,13 +83,12 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
...
             'stimulus_id': ('presentation', [0] if isinstance(text, str) else np.arange(len(text))),
         }
+
         if self.behavioral_task:
-            logits = base_output.logits
-            pred_id = torch.argmax(logits, axis=2).squeeze()
-            last_model_token_inference = pred_id[-1].tolist()
-            next_word = self.tokenizer.decode(last_model_token_inference)
+            behavioral_output = self.run_behavioral_experiment(text=text,
+                                                    base_output= base_output)
```
Comment (mschrimpf) [inline brainscore_language/models/huggingface.py:89]: As above (+ `text` isn't actually used)

```suggestion
            behavioral_output = self.output_to_behavior(base_output= base_output)
```

File: brainscore_language/models/huggingface.py [hunk 3]
```diff
@@ -112,8 +117,18 @@ def digest_text(self, text: Union[str, List[str]]) -> Dict[str, DataAssembly]:
                 coords={**stimuli_coords, **neuroid_coords},
                 dims=['presentation', 'neuroid'])
             output['neural'] = representations
+
         return output
 
+    def predict_next_word(self, text, base_output):
+
```
Comment (mschrimpf) [inline brainscore_language/models/huggingface.py:124]:
```suggestion
    def predict_next_word(self, base_output):
```

File: tests/test_models/test_huggingface.py [hunk 4]
```diff
@@ -1,13 +1,57 @@
...
+    #     """
+    #
+    #     model = HuggingfaceSubject(model_id='gpt2-xl',
+    #                                region_layer_mapping={}
+    #                                 )
+    #     text = 'the quick brown fox'
+    #     logging.info(f'Running {model.identifier()} with text "{text}"')
+    #     model.perform_behavioral_task(task=ArtificialSubject.Task.next_word)
+    #     next_word = model.digest_text(text)['behavior'].values
+    #     assert next_word == ' jumps'
```
Comment (mschrimpf) [inline tests/test_models/test_huggingface.py:51]: dead code?

File: tests/test_models/test_huggingface.py [hunk 5]
```diff
@@ -1,13 +1,57 @@
...
+    # @pytest.mark.memory_intense
+    # def test_next_word_mask_bert_base_uncased(self):
+    #     """
+    #     This is a simple test that takes in text = 'the quick brown fox', and asserts
+    #     that the next word predicted is '.'
+    #     This test is a stand-in prototype to check if our model definitions are correct.
+    #     """
+    #     model = HuggingfaceSubject(model_id='bert-base-uncased',
+    #                                 # model_class=AutoModelForCausalLM,
+    #                                 # tokenizer_class=AutoTokenizer,
```
Comment (mschrimpf) [inline tests/test_models/test_huggingface.py:24]: extra dead code

File: tests/test_models/test_huggingface.py (hunk 4 above)
Comment (d-val) [reply tests/test_models/test_huggingface.py]: I was commenting in and out tests as tests. All done, and uncommented.

File: brainscore_language/models/huggingface.py (hunk 1 above)
Comment (d-val) [reply brainscore_language/models/huggingface.py:46]: done!

File: brainscore_language/models/huggingface.py (hunk 3 above)
Comment (d-val) [reply brainscore_language/models/huggingface.py:124]: done!

File: tests/test_models/test_huggingface.py [hunk 6]
```diff
@@ -1,13 +1,57 @@
...
+    @pytest.mark.memory_intense
+    def test_next_word_mask_bert_base_uncased(self):
+        """
+        This is a simple test that takes in text = 'the quick brown fox', and asserts
+        that the next word predicted is '.'
+        This test is a stand-in prototype to check if our model definitions are correct.
+        """
+        model = HuggingfaceSubject(model_id='bert-base-uncased',
+                                    # model_class=AutoModelForCausalLM,
+                                    # tokenizer_class=AutoTokenizer,
```
Comment (mschrimpf) [inline tests/test_models/test_huggingface.py:24]: dead code, remove
```suggestion
```

**Review by mschrimpf — APPROVED** [review]
one instance of dead code, fine to merge after that is fixed!

**Review by kvfairchild — APPROVED** [review]
